core/src/ldml/ldml_processor.cpp

```diff
@@ -59,23 +59,130 @@ namespace {
+  const kmx::COMP_KMXPLUS_STRS *strs = nullptr;
+  const kmx::COMP_KMXPLUS_KEYS *keys = nullptr;
+
   // TODO-LDML: load the file from the buffer (KMXPlus format)
   // Note: kb_path is essentially debug metadata here
-  assert(data.size() != 0);
+  KMXPLUS_ASSERT(true, data.size() > sizeof(kmx::COMP_KEYBOARD_EX));
 
+  // Locate the structs here, but still retain ptrs to the raw structs.
   const kmx::PCOMP_KEYBOARD comp_keyboard = (kmx::PCOMP_KEYBOARD)data.data();
+  KMXPLUS_ASSERT(true, !!(comp_keyboard->dwFlags & KF_KMXPLUS));
+  const kmx::COMP_KEYBOARD_EX* ex = reinterpret_cast<const kmx::COMP_KEYBOARD_EX*>(comp_keyboard);
+
+  // validate size and offset
+  KMXPLUS_ASSERT(true, ex->kmxplus.dwKMXPlusSize >  LDML_LENGTH_SECT);
+  KMXPLUS_ASSERT(true, ex->kmxplus.dwKMXPlusSize <  (data.size() - sizeof(kmx::COMP_KEYBOARD_EX)));
+  KMXPLUS_ASSERT(true, ex->kmxplus.dpKMXPlus     >= sizeof(kmx::COMP_KEYBOARD_EX));
+  KMXPLUS_ASSERT(true, ex->kmxplus.dpKMXPlus     <  data.size());
+
+  // calculate pointers to start and end of kmxplus
+  const uint8_t * const kmxplusdata   = data.data() + ex->kmxplus.dpKMXPlus;      // Start of + data
+  // const uint8_t* kmxpluslimit  = kmxplusdata    + ex->kmxplus.dwKMXPlusSize;  // End of   + data
+
+  // Now load sections.
+
+  // This will validate (and possibly print) all data.
+  // TODO-LDML: we will be replacing this validation with validation-as-we-go below.
+  KMXPLUS_ASSERT(true, kmx::validate_kmxplus_data(kmxplusdata));
+
+  // Get out the SECT header
+  {
+    const KMX_DWORD offset = 0;
+    sect = kmx::as_kmxplus_sect(kmxplusdata+offset);
+    KMXPLUS_ASSERT(true, sect != nullptr);
+    // Specified data size fits in total
+    KMXPLUS_ASSERT(true, (offset+sect->header.size) < ex->kmxplus.dwKMXPlusSize);
+  }
 
-  if (validate_kmxplus_data(comp_keyboard)) {
-    _valid = true;
+  // Fill out the other sections we need.
+  {
+    const KMX_DWORD offset = sect->find(LDML_SECTIONID_STRS);
+    KMXPLUS_ASSERT(true, offset != 0);
+    KMXPLUS_ASSERT(true, offset < ex->kmxplus.dwKMXPlusSize);
+    strs = kmx::as_kmxplus_strs(kmxplusdata+offset);
+    KMXPLUS_ASSERT(true, strs != nullptr);
+    // Specified data size fits in total
+    KMXPLUS_ASSERT(true, (offset+strs->header.size) < ex->kmxplus.dwKMXPlusSize);
   }
+
+  {
+    const KMX_DWORD offset = sect->find(LDML_SECTIONID_KEYS);
+    KMXPLUS_ASSERT(true, offset != 0);
+    keys = kmx::as_kmxplus_keys(kmxplusdata+offset);
+    KMXPLUS_ASSERT(true, keys != nullptr);
+    // Specified data size fits in total
+    KMXPLUS_ASSERT(true, (offset+keys->header.size) < ex->kmxplus.dwKMXPlusSize);
+
+    // read all keys into array
+    for (KMX_DWORD i=0; i<keys->count; i++) {
+      const kmx::COMP_KMXPLUS_KEYS_ENTRY &entry = keys->entries[i];
+      KMX_DWORD len = 0;
+      KMX_WCHAR out[BUFSIZ];
+      if (entry.flags && LDML_KEYS_FLAGS_EXTEND) {
+        KMXPLUS_ASSERT(false, nullptr == strs->get(entry.to, out, BUFSIZ));
```

Comment:
I think this may have been asked but I missed the answer. Is the reason to just use a static buffer of size 512  because we know it will always be long enough?  Otherwise, we could make a get accept a nullptr for the buffer and return the required size.

Reply:
this is going to change to using a string buffer, https://github.com/keymanapp/keyman/issues/7134

Reply:
I'm going to do the string work in a separate PR, concentrating on validation here.